pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiutil
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"time"
+
+	"golang.org/x/time/rate"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// ErrRateLimited is returned by a dynamicRESTMapper method if the number
+// of API calls has exceeded a limit within a certain time period.
+type ErrRateLimited struct {
+	// Duration to wait until the next API call can be made.
+	Delay time.Duration
+}
+
+const errRLMsg = "too many API calls to the dynamicRESTMapper within a timeframe"
+
+func (e ErrRateLimited) Error() string {
+	return fmt.Sprintf("%s (%dns)", errRLMsg, int64(e.Delay))
+}
+
+var errRLRe = regexp.MustCompile(fmt.Sprintf(".*%s \\(([0-9]+)ns\\).*", errRLMsg))
+
+func IsRateLimited(err error) (time.Duration, bool) {
+	if e, ok := err.(ErrRateLimited); ok {
+		return e.Delay, true
+	}
+	if matches := errRLRe.FindStringSubmatch(err.Error()); len(matches) > 1 {
+		d, err := strconv.Atoi(matches[1])
+		if err == nil {
+			return time.Duration(d), true
+		}
+	}
+	return 0, false
+}
+
+var (
+	// LimitRate is the number of dynamicRESTMapper API calls allowed per second
+	// assuming the rate of API calls <= LimitRate.
+	// There is likely no need to change the default value.
+	LimitRate = 600
+	// LimitSize is the maximum number of simultaneous dynamicRESTMapper API
+	// calls allowed.
+	// There is likely no need to change the default value.
+	LimitSize = 5
+)
+
+// dynamicRESTMapper is a RESTMapper that dynamically discovers resource
+// types at runtime. This is in contrast to controller-manager's default
+// RESTMapper, which only checks resource types at startup, and so can't
+// handle the case of first creating a CRD and then creating an instance
+// of that CRD.
+type dynamicRESTMapper struct {
+	client       discovery.DiscoveryInterface
+	staticMapper meta.RESTMapper
+	limiter      *dynamicLimiter
+}
+
+// NewDynamicRESTMapper returns a dynamicRESTMapper for cfg.
+func NewDynamicRESTMapper(cfg *rest.Config) (meta.RESTMapper, error) {
```

Comment:
Not sure if controller-runtime has a convention for constructor design, but to handle the configurable per-rest-mapper limiter, would the functional options pattern make sense here?

```go
func WithLimiter(limiter *rate.Limiter) func(*dynamicRESTMapper) error {
    return func(drm *dynamicRESTMapper) error {
        drm.limiter = limiter
        return nil
    }
}

func NewDynamicRESTMapper(cfg *rest.Config, opts ...func(drm *dynamicRESTMapper) error) (meta.RESTMapper, error) {
    // setup the default dynamic restmapper, with a default limiter
    drm := ...

    for _, opt := range opts {
        if err := opt(drm); err != nil {
           return nil, err
        }
    }
    ...
}
```

This would give us 3 things:
1. Per-rest mapper configurable limiter
2. Callers could still use `NewDynamicRESTMapper(restConfig)` to use the default, or override with `NewDynamicRESTMapper(restConfig, WithLimiter(customLimiter))`
3. If new configurations are needed later, they could be added without changing the constructor signature.

Reply:
So, we tend not to favor constructors unless absolutely necessary (prefering sync.Once defaulting), but this seemed like a place where just having the constructor could be useful.  In that case, functional options are fine.